lib/cf-h3-proxy.c

```diff
@@ -0,0 +1,3038 @@
+  x.data = data;
+  x.s = s;
+  while(s->ssl && !s->closed && !eagain &&
+         (total_recv_len < PROXY_H3_STREAM_CHUNK_SIZE)) {
+
+    if(proxy_ctx->partial_read && s->id == proxy_ctx->tunnel.stream_id) {
+      handle_buffered_data(cf, data);
+      break;
+    }
+
+    if(Curl_bufq_is_empty(&s->recvbuf) && !s->recvd_eos) {
+      while(!eagain && !s->recvd_eos && !Curl_bufq_is_full(&s->recvbuf)) {
+        result = Curl_bufq_sipn(&s->recvbuf, 0, h3_quic_recv, &x, &n);
+        if(result) {
+          if(result != CURLE_AGAIN) {
+            goto out;
+          }
+          result = CURLE_OK;
+          eagain = TRUE;
+        }
+      }
+    }
+
+    /* At this point we can have 2 scenarios:
+      (1) The proxytunnel is NOT yet UP and we are still negotiating the
+          CONNECT request and the different unidirectional streams with
+          the proxy. In this case, all the data must be forwarded to
+          nghttp3 library for processing.
+          Flow:
+          cf_h3_proxy_quic_connect() --> proxy_h3_submit() -->
+          proxy_h3_progress_egress() --> proxy_h3_progress_ingress() -->
+          cf_osslq_h3conn_add_stream() --> cf_osslq_stream_recv() -->
+          inspect_response() --> tunnel is UP (with bidi stream id = 0)
+      (2) The proxytunnel is UP
+          At this point, we have 7 streams - 1 bidi (the tunnel stream) and
+          6 unidirectional streams (3 from curl and 3 from the proxy)
+          Every "DATA" from the underlying HTTP/1.1 connection must be
+          forwarded end-to-end through this HTTP/3 proxytunnel
+          (i) Stream 0:
+          <HTTP/1.1 data> === proxytunnel === <HTTP/3 headers>
+                                              <HTTP/3 data> = <HTTP/1.1 data>
+          (ii) Unidirectional Streams:
+          can be terminated here, HTTP/1.1 layer is unaware of this
+          Functions of Interest:
+          (1) nghttp3_conn_read_stream --> this received HTTP/3 specific info
+          (2) cb_h3_recv_data --> this received the actual end-to-end flow
+                                      data from the server via the proxy
+          nghttp3_conn_read_stream() internally invokes cb_h3_recv_data()
+          In cb_h3_recv_data(), we are storing the "data" received w.r.t. to
+          the HTTP/1.1 flow in cf_h3_proxy_ctx->inbufq
+          Now, we need to propagate this up to the recv() call from the
+          HTTP/1.1 SSL layer
+          This is how the filter chain looks like:
+          Curl_cft_http_connect --> Curl_cft_ssl --> ... --> Curl_cft_h3_proxy
+    */
+
+    /* Forward what we have to nghttp3 */
+    if(!Curl_bufq_is_empty(&s->recvbuf)) {
+      const unsigned char *buf;
+      size_t blen;
+
+      while(Curl_bufq_peek(&s->recvbuf, &buf, &blen)) {
+        nread = nghttp3_conn_read_stream(ctx->h3.conn, s->id,
+                                         buf, blen, 0);
+        if(nread < 0) {
+          failf(data, "nghttp3_conn_read_stream(len=%zu) error: %s",
```

Comment:
As it sets result to CURLE_OK below here, should this perhaps be a plain infof() instead?